getnacelle/nacelle-js
# feat: scaffold commerce queries plugin

## Why are these changes introduced?

Addresses [ENG-8179](https://nacelle.atlassian.net/browse/ENG-8179) 



> Setup plugin for storefront client methods

## What is this pull request doing?



### `@nacelle/storefront-sdk-plugin-commerce-queries`

This PR scaffolds a new Storefront SDK plugin, `@nacelle/storefront-sdk-plugin-commerce-queries`. The purpose of this plugin is to support `@nacelle/storefront-sdk@1.x` users who rely on methods like `.products`, `.content` etc. in their projects. By moving these methods to a plugin, we allow the `@nacelle/storefront-sdk` to be free of KBs of JavaScript related to those REST-y methods, which will benefit users who use `client.query` exclusively.

The scaffolding includes `package.json` configuration, TypeScript configuration, GraphQL Codegen configuration, and Vitest setup.

### `@nacelle/storefront-sdk`

This PR updates `packages/storefront-sdk/package.json` according to [TypeScript's docs on ECMAScript Modules in Node.js](https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing) to support native ESM projects.

Other changes includes:
- The remaining TypeScript Operations code generation has been moved to `__mocks__`.
- Because there are now two generated types files, this PR breaks the `codegen` script into multiple pieces. Now, `codegen` only generates the types files, `codegen:format` applies Prettier formatting, `codegen:lint` applies ESLint, and `codegen:all` does generation, formatting, and linting.

### Monorepo-wide

This PR adjusts the Lerna config to accommodate subdirectories in `packages/`.

## How to Test



1. Check out the `ENG-8396-setup-plugin-for-storefront-client-methods` branch.
2. `npm run bootstrap`
3. Navigate to `packages/storefront-sdk`
4. `npm run codegen:all` - there should be no uncommitted changes
5. `npm run build` - the build should not fail
6. `npm run coverage` - all tests should pass with 100% coverage
7. Navigate to `packages/storefront-sdk-plugins/commerce-queries`
8. `npm run codegen:all` - there should be no uncommitted changes
9. `npm run build` - the build should not fail
10. `npm run coverage` - all tests should pass

## Checklist


[ENG-8179]: https://nacelle.atlassian.net/browse/ENG-8179?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Callouts and next steps

The TypeScript config in `packages/storefront-sdk-plugins/commerce-queries/tsconfig.json` is fragile and problematic. The hope is that after `@nacelle/storefront-sdk@2.0.0-beta.1` is published, we can do away with:
1. `"baseUrl": ".",`
2. The `"paths"` config
3. [`"preserveSymlinks": true,`](https://www.typescriptlang.org/tsconfig#preserveSymlinks)
